### ISSUE TITLE
Broken Samples

### ISSUE THREAD

**Issue by jwoertink** [ysbaddaden/sdl.cr#3]
I've gone through all the files in the samples folder, and some do work fine, but there are some that don't. Most of the errors are the same, so might be a small fix (hopefully) to get them all running. 
These are the errors I get running on Crystal 0.20.3 on macOS 10.12.2

```
[15:52PM] samples (master)$ crystal 04_keyboard.cr 
SDL_GetWindowSurface:  (SDL::Error)
0x10c547562: *CallStack::unwind:Array(Pointer(Void)) at ??
0x10c547501: *CallStack#initialize:Array(Pointer(Void)) at ??
0x10c5474d8: *CallStack::new:CallStack at ??
0x10c546ae1: *raise<SDL::Error>:NoReturn at ??
0x10c572899: *SDL::Window#surface:SDL::Surface at ??
0x10c5425d8: __crystal_main at ??
0x10c5465b8: main at ??
```
```
[15:52PM] samples (master)$ crystal 05_stretching.cr 
SDL_GetWindowSurface:  (SDL::Error)
0x104699262: *CallStack::unwind:Array(Pointer(Void)) at ??
0x104699201: *CallStack#initialize:Array(Pointer(Void)) at ??
0x1046991d8: *CallStack::new:CallStack at ??
0x104698781: *raise<SDL::Error>:NoReturn at ??
0x1046c4599: *SDL::Window#surface:SDL::Surface at ??
0x104698880: *load_bmp<String, SDL::Window>:SDL::Surface at ??
0x104694002: __crystal_main at ??
0x104698258: main at ??
```
```
[15:52PM] samples (master)$ crystal 06_image.cr 
libpng warning: iCCP: known incorrect sRGB profile
SDL_GetWindowSurface:  (SDL::Error)
0x10f692002: *CallStack::unwind:Array(Pointer(Void)) at ??
0x10f691fa1: *CallStack#initialize:Array(Pointer(Void)) at ??
0x10f691f78: *CallStack::new:CallStack at ??
0x10f691561: *raise<SDL::Error>:NoReturn at ??
0x10f6bd289: *SDL::Window#surface:SDL::Surface at ??
0x10f68cebf: __crystal_main at ??
0x10f691038: main at ??
```
```
[15:58PM] samples (master)$ crystal 14_animated_sprites_and_vsync.cr 
SDL_GetWindowSurface: No hardware accelerated renderers available (SDL::Error)
0x10e600ef2: *CallStack::unwind:Array(Pointer(Void)) at ??
0x10e600e91: *CallStack#initialize:Array(Pointer(Void)) at ??
0x10e600e68: *CallStack::new:CallStack at ??
0x10e600411: *raise<SDL::Error>:NoReturn at ??
0x10e62c6b9: *SDL::Window#surface:SDL::Surface at ??
0x10e5fbda1: __crystal_main at ??
0x10e5ffee8: main at ??
```
```
[16:10PM] samples (master)$ crystal 15_rotation_and_flipping.cr 
libpng warning: iCCP: known incorrect sRGB profile
SDL_GetWindowSurface: No hardware accelerated renderers available (SDL::Error)
0x101f30342: *CallStack::unwind:Array(Pointer(Void)) at ??
0x101f302e1: *CallStack#initialize:Array(Pointer(Void)) at ??
0x101f302b8: *CallStack::new:CallStack at ??
0x101f2f861: *raise<SDL::Error>:NoReturn at ??
0x101f5ba89: *SDL::Window#surface:SDL::Surface at ??
0x101f2b234: __crystal_main at ??
0x101f2f338: main at ??
```
```
[16:10PM] samples (master)$ crystal basic.cr 
SDL_GetWindowSurface:  (SDL::Error)
0x100343302: *CallStack::unwind:Array(Pointer(Void)) at ??
0x1003432a1: *CallStack#initialize:Array(Pointer(Void)) at ??
0x100343278: *CallStack::new:CallStack at ??
0x1003428a1: *raise<SDL::Error>:NoReturn at ??
0x10036dca9: *SDL::Window#surface:SDL::Surface at ??
0x10033e38b: __crystal_main at ??
0x100342378: main at ??
```
```
[16:10PM] samples (master)$ crystal bmp.cr 
SDL_GetWindowSurface:  (SDL::Error)
0x103127722: *CallStack::unwind:Array(Pointer(Void)) at ??
0x1031276c1: *CallStack#initialize:Array(Pointer(Void)) at ??
0x103127698: *CallStack::new:CallStack at ??
0x103126ca1: *raise<SDL::Error>:NoReturn at ??
0x103152a59: *SDL::Window#surface:SDL::Surface at ??
0x10312279e: __crystal_main at ??
0x103126778: main at ??
```

The samples 7, 8, 9, 10, 11, 12, and 13 all work as expected. I did confirm the original C++ samples run when built with XCode.

**Comment by ysbaddaden** [comment]
Yes, I just noticed a stupid issue. Please try the latest master or at least 42bc2c84fc716c07e76a29e52886f0b2fc9c22b6 (the latest commit requires Crystal 0.20.4). This should fix the empty message exceptions (but not 14 and 15).

**Comment by jwoertink** [comment]
oh sweet! Ok, upgraded to 0.20.4, and pulled latest master. All samples work except 14 and 15 now. I'll get a warning `libpng warning: iCCP: known incorrect sRGB profile`, but I don't think that's a huge issue. 

The errors from 14 & 15 are the same

```
[16:37PM] samples (master)$ crystal 14_animated_sprites_and_vsync.cr 
SDL_GetWindowSurface: No hardware accelerated renderers available (SDL::Error)
0x103db2c42: *CallStack::unwind:Array(Pointer(Void)) at ??
0x103db2be1: *CallStack#initialize:Array(Pointer(Void)) at ??
0x103db2bb8: *CallStack::new:CallStack at ??
0x103db2141: *raise<SDL::Error>:NoReturn at ??
0x103dde405: *SDL::Window#surface:SDL::Surface at ??
0x103dadab9: __crystal_main at ??
0x103db1c18: main at ??
[16:37PM] samples (master)$ crystal 15_rotation_and_flipping.cr 
libpng warning: iCCP: known incorrect sRGB profile
SDL_GetWindowSurface: No hardware accelerated renderers available (SDL::Error)
0x103cef092: *CallStack::unwind:Array(Pointer(Void)) at ??
0x103cef031: *CallStack#initialize:Array(Pointer(Void)) at ??
0x103cef008: *CallStack::new:CallStack at ??
0x103cee591: *raise<SDL::Error>:NoReturn at ??
0x103d1a7d5: *SDL::Window#surface:SDL::Surface at ??
0x103ce9f4c: __crystal_main at ??
0x103cee068: main at ??
```

**Comment by jwoertink** [comment]
On a side note, do you want those sample images in the repo?

**Comment by ysbaddaden** [comment]
No, I didn't put images because that would only clutter the repository, and installing the shard dependency would always install the images, too...

I found a reference to the issue, on a French forum: https://openclassrooms.com/forum/sujet/sdl2-fillrect-donne-une-segfault

The problem seems to be that I call `window.surface` which is a software renderer, but we created a hardware accelerated renderer before, so it fails. Though SDL doesn't care on linux :/

**Comment by ysbaddaden** [comment]
Maybe fixed in c7a46c6?

**Comment by jwoertink** [comment]
Yup! That was it! 🎉  Thanks for the help!!
